core/domain/expression_parser.py

```diff
@@ -113,14 +125,29 @@ def tokenize(expression):
             expression.
 
     Raises:
-        Exception: Invalid syntax: Unexpected token present in the expression.
+        Exception: Invalid token.
     """
     expression = expression.replace(' ', '')
 
-    re_string = r'([a-zA-Z]+|[0-9]+\.[0-9]+|[0-9]+|[%s])' % '\\'.join(
-        _VALID_OPERATORS)
+    # Note: Greek letters and math functions are sorted in reverse by length so
+    # that longer ones get matched before shorter ones.
+    # For eg. 'x + epsilon' should be tokenized as ['x','+','epsilon'] and not
+    # ['x','+','e','*','psi','*','l','*','o','*','n']. a^2.
+    re_string = r'(%s|[a-zA-Z]|[0-9]+\.[0-9]+|[0-9]+|[%s])' % (
+        '|'.join(sorted(
+            _GREEK_LETTERS + _MATH_FUNCTION_NAMES, reverse=True, key=len)),
+        '\\'.join(_VALID_OPERATORS))
 
     token_texts = re.findall(re_string, expression)
+
+    original_exp_frequency = collections.Counter(expression)
+    tokenized_exp_frequency = collections.Counter(''.join(token_texts))
+
+    for character in original_exp_frequency:
```

Comment:
Great check, though why don't you just join token_texts and compare the resulting string with "expression" (i.e. why do you need the counter)?

Reply:
In the faulty scenario, there will be one or more characters that are present in the original expression but not in the tokenized expression (because the regex might've skipped them due to them being invalid). But this doesn't mean that if one instance of a character gets skipped, all of them will. For eg. in `1.2 + 3.4.5`, only the last period (`.`), will be skipped so we need to compare the frequency. Alternatively, we could sort both expressions and then compare with a single pass but this seems simpler, plus this has an amortized linear time complexity and sorting would have logarithmic (not that it matters too much in this case since the expressions aren't going to be too long, but still, something to consider).

Reply:
Sorry, I'm not sure I follow. What I meant was, if x = `1.2 + 3.4.5` gets tokenized as t = [`1.2`, `+`, `3.4`], then why not just do `x == ''.join(t)`?

Reply:
No -- `1.2 + 3.4.5` gets tokenized as `[1.2, +, 3.4, 5]`. If I join that it's `1.2+3.45` which is not equal to `1.2+3.4.5` so we can check it's validity, but we won't know which character is the invalid one which is why we need to count the frequency which will then suggest that the frequency of `.` is supposed to be 3 but is actually 2 in the tokenized expression and so we can say that `.` is the invalid/unexpected character.

Reply:
I see, makes sense. OK, thanks for explaining; LGTM!

In a future PR please add a comment here to explain that you're using a counter so that you can figure out the right character to include in the error message.

Reply:
Will do, thanks!